functional_tests/README.md

```diff
@@ -0,0 +1,147 @@
+Docker Container Functional Tests
+========================================
+
+This code builds an image that serves as a framework to run functional tests on
+a target image. The image under tests runs as a docker container inside of this
+image, which itself runs as a docker container on a host machine in the cloud
+(when run through a
+[Google Cloud Container Build](https://cloud.google.com/container-builder/docs/overview)
+).
+
+# How Use This Image
+
+This image can be first built using the included `cloudbuild.yaml`. The
+following will build a `functional_test` GCR image in your project.
+
+``` shell
+gcloud container builds submit --config cloudbuild.yaml .
+```
+
+Then, you can run another container build using this image to execute the test.
+An example is given under `examples/simple/cloudbuild.yaml`, which assumes
+the `functional_test` image is in the same project, and will run the tests
+specified in `test.yaml`. You can run this example as followed:
+
+``` shell
+cd examples/simple
+gcloud container builds submit --config cloudbuild.yaml .
+```
+
+Essentially, `test.yaml` is specified as an argument to `functional_test`:
+
+``` yaml
+steps:
+- name: gcr.io/$PROJECT_ID/functional_test
+  args: [--test_spec, test.yaml]
+```
+
+# Test Specifications
+
+The tests are specified by a YAML or JSON file. Currently, this image can only
+execute tests from one file at a time.
+
+## Setup
+
+The setup section of the file specifies commands to setup a running container
+to test. These commands run once for all tests.
+
+This section can spin up multiple containers should the test subject need
+dependencies.
+
+For example, the following starts a container named `some-redis` running a redis
+server and waits 5 seconds to make sure the server runs.
+
+``` yaml
+setup:
+- command: [docker, run, --name, some-redis, -d, launcher.gcr.io/google/redis3]
```

Comment:
One thought here, but feel free to ignore if you don't like the idea...

It feels like the only piece of information the user needs to specify here is the image:
setup:
  image: launcher.gcr.io/google/redis3
  
The name of the container could be random (would eliminate "target" below).

If there is a need to customize the command line, then this could be extended to:
setup:
  image: ...
  options: ...
  command: ...
  args: [...]

Which results in a docker run command like:
$ docker run <options> <image> <command> <args>